core/shared_obj.h

```diff
@@ -0,0 +1,133 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_SHARED_OBJ_H_
+#define BESS_SHARED_OBJ_H_
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+
+#include "utils/common.h"  // for PairHasher functor
+
+namespace bess {
+
+// SharedObjectSpace provides a simple mechanism for independent modules to
+// share arbitrary objects. There is a global "shared_objects" instance in the
+// "bess" namespace (see shared_obj.cc). Modules (or port drivers) can use the
+// global instance to create and access shared objects by name. Shared objects
+// are instances of an arbitrary class T, referenced by a shared_ptr<T>. Just
+// like any other shared_ptr objects, shared objects are automatically
+// destructed once all references to the object have gone.
+//
+// Usage:
+// shared_ptr<TypeFoo> bess::shared_objects::Get<TypeFoo>("foo_name");
+//
+// If there is no object named "foo_name", a new instance of TypeFoo is created
+// (of course with its constructor). The module is expected to keep the
+// shared_ptr to the object until it is no longer needed.
+//
+// Type safety: The type of an object must be identical for all users of the
+// object. To prevent type errors, this class provides a separate namespace
+// for each object type. For example, if another module requests an object
+// also named "foo_name" but with a different type other than TypeFoo, a
+// different object will be returned.
+//
+// Thread safety: Get() and Lookup() functions are thread safe as they are
+// protected by a mutex. However, shared objects themselves are not protected
+// by default; you should use any synchronization mechanism for objects as
+// necessary.
+class SharedObjectSpace {
+ public:
+  // By default, the Get() method creates a new object with its default
+  // constructor (a constructor without any arguments). To change this behavior,
+  // you can use the optional argument "constructor", a callable object
+  // (function pointer, functor, lambda function, etc.) that creates the object
+  // in a way you'd like, e.g., to create an object with a non-default
+  // constructor, or to reuse an already existing object.
```

Comment:
Nice.  Should perhaps warn users that this holds the lock during the creation, though, so their creator should not block on anything that might also require a shared creation.

Reply:
Good point. I'll add a comment here.

Reply:
* Renamed "constructor" to your term "creator" to avoid confusion with object constructor.
* Realized the creator might want to create another shared object. Replaced `mutex` with `recursive_mutex`.